tests/test_divergence_free_dist.py

```diff
@@ -0,0 +1,118 @@
+
+
+def rew_fn_1_potential_shaping(state, action, next_state, _):
+    return state + next_state + action + next_state / 3 - state / 3
+
+
+def rew_fn_2(state, action, next_state, _):
+    return state**2 + next_state**2 + 2 * action
+
+
+def test_divergence_free_dist_no_errors():
+    state_space = gym.spaces.Discrete(10)
+    action_space = gym.spaces.Discrete(10)
+
+    x = rew_fn_1
+    y = rew_fn_2
+
+    dist = divergence_free.DivergenceFree(
+        state_sampler=DummyGymStateSampler(space=state_space),
+        action_sampler=GymSpaceSampler(space=action_space),
+        discount_factor=1,
+    ).distance(x, y, n_samples_cov=500, n_samples_can=1000)
+
+    assert isinstance(dist, float)
+    assert not np.isnan(dist)
+
+
+def test_divergence_free_dist_reward_equivalence_constant_reward():
+    state_space = gym.spaces.Discrete(10)
+    action_space = gym.spaces.Discrete(10)
+
+    x = rew_fn_0
+    y = rew_fn_0_potential_shaping
+
+    dist = divergence_free.DivergenceFree(
+        state_sampler=DummyGymStateSampler(space=state_space),
+        action_sampler=GymSpaceSampler(space=action_space),
+        discount_factor=1,
+    ).distance(x, y, n_samples_cov=500, n_samples_can=2500)
+
+    print(dist)
+
+    assert np.isclose(dist, 0, atol=1e-7)
+
+
+def test_divergence_free_dist_reward_equivalence_constant_reward_multiple_dims():
+    state_space = gym.spaces.MultiDiscrete([10, 10])
+    action_space = gym.spaces.MultiDiscrete([10, 10])
+
+    x = rew_fn_0
+    y = rew_fn_0_potential_shaping
+
+    dist = divergence_free.DivergenceFree(
+        state_sampler=DummyGymStateSampler(space=state_space),
+        action_sampler=GymSpaceSampler(space=action_space),
+        discount_factor=1,
+    ).distance(x, y, n_samples_cov=500, n_samples_can=2500)
+
+    print(dist)
+
+    assert np.isclose(dist, 0, atol=1e-7)
+
+
+def test_divergence_free_dist_reward_equivalence_constant_reward_multiple_dims_continuous():
+    state_space = gym.spaces.Box(low=-10, high=10, shape=(10,))
+    action_space = gym.spaces.Box(low=-10, high=10, shape=(10,))
+
+    x = rew_fn_0
+    y = rew_fn_0_potential_shaping
+
+    dist = divergence_free.DivergenceFree(
+        state_sampler=DummyGymStateSampler(space=state_space),
+        action_sampler=GymSpaceSampler(space=action_space),
+        discount_factor=1,
+    ).distance(x, y, n_samples_cov=500, n_samples_can=10000)
+
+    print(dist)
+
+    assert np.isclose(dist, 0, atol=1e-7)
+
+
+def test_divergence_free_dist_reward_equivalence_linear_reward():
+    state_space = gym.spaces.Discrete(10)
+    action_space = gym.spaces.Discrete(10)
+
+    x = rew_fn_1
+    y = rew_fn_1_potential_shaping
+
+    dist = divergence_free.DivergenceFree(
+        state_sampler=DummyGymStateSampler(space=state_space),
+        action_sampler=GymSpaceSampler(space=action_space),
+        discount_factor=1,
+    ).distance(x, y, n_samples_cov=500, n_samples_can=5000)
+
+    print(dist)
+
+    assert np.isclose(dist, 0, atol=2e-1)
```

Comment:
I expected once things are working well, we should get significantly lower distance than 0.2 for linear potentials, lmk if that seems hard to achieve

Reply:
I think I've been getting somewhere between 0.05 and 0.2 (most runs below 0.1). I can maybe try increasing the depth of the network if it seems like it's definitely possible to do better.

Reply:
Well, in some sense it's definitely possible to do better: if our function class contains good approximations to any linear function, in particular to the potential, we should be able to get very close to zero distance. That said, I'm not sure how easy this is in practice (i.e. with imperfect optimization). If it turns out we can't get lower than this in such a simple setting though, that's bad news for the entire method. So I think worth figuring out what the limits are here before moving to more complicated cases.

One sanity check: if you use a linear model for the potential (i.e. just one `nn.Linear`), then I think you should be able to get very low distances (`< 1e-2` hopefully). If that's not working, we should look into that first, rather than scaling up the network (and in general, it seems somewhat weird if we'd need to scale up the network just to fit a linear function)

Reply:
Yeah that works (0.005 distance) (if I also increase `n_can_samples` to 25k). I'll try increasing the number of samples and go back to the MLP net to see if I can get it to work.

Reply:
I can get it to `< 5e-2` now, but going much lower seems like it's gonna be pretty difficult. I simplified to only use two linear layers with a gelu nonlinearity between them and also dramatically increased the number of canonicalization samples.

Reply:
Ok, I was hoping this would work even better but sounds plausible that's hard. Probably makes sense to try more complicated potentials and see how things scale. If we can get to 5e-2 even in more complex settings, that's probably fine, whereas if things get much worse, we might have to come up with new ideas.

Two other things that might be good to test:
- How stable is optimization? I.e. if you train several times (with different optimization), do you get basically the same distance each time, or does it fluctuate a lot? This might be most visible in more complicated settings, and with reward functions that don't just differ by potential shaping.
- What if you canonicalize the zero reward function? In that case, the network can clearly fit the correct potential (by just setting all weights to zero). If this doesn't work any better than the `5e-2`, it indicates we might want to push a bit more after all (even more optimization)

Reply:
- It seems reasonably stable, I'd say the fluctuations have been around the `1e-2` mark (Currently seems to be like `2.5e-2 +- 1e-2`. I'll run more extensive tests tomorrow to get a better idea of the variance in outcomes.
- Canonicalizing the zero reward function works, you can see other tests in the file that check for that in a few different cases (it just gets a distance of 0, since the samples are low variance enough that we treat both the zero reward function and the zero reward function with the potential as constant reward functions).

Reply:
Ok, if I read the code correctly that means the zero reward functions gets canonicalized to variance less than `1e-2`, which seems pretty reasonable. If you do spend more time on improving the optimization, you could check just how far away the canonicalized reward is from exactly 0 (i.e. is the variance more like `5e-3` or `1e-5`). That seems like a good simple test case to tell whether you're making progress. But also very reasonable at this point to just scale things up and see what happens (and perhaps eventually move towards a pipeline with actual environments and reward models)